platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java

```diff
@@ -812,7 +816,31 @@ public void addOnRemoveUnusedStyleImagesListener(@NonNull OnRemoveUnusedStyleIma
    * @param listener The callback that's invoked when map needs to release unused image resources
    */
   public void removeOnRemoveUnusedStyleImagesListener(@NonNull OnRemoveUnusedStyleImagesListener listener) {
-    mapChangeReceiver.removeOnRemoveUnusedStyleImagesListener(listener);
+    boolean lastListenerRemoved = mapChangeReceiver.removeOnRemoveUnusedStyleImagesListener(listener);
+    if (lastListenerRemoved) {
+      addDefaultRemoveUnusedStyleImagesListener();
+    }
+  }
+
+  private void addDefaultRemoveUnusedStyleImagesListener() {
+    if (defaultOnRemoveUnusedStyleImagesListener == null) {
+      // Initialize default callback for OnRemoveUnusedStyleImages notifications.
+      defaultOnRemoveUnusedStyleImagesListener = new OnRemoveUnusedStyleImagesListener() {
```

Comment:
Do you think it would be worth providing the default handler on core level? i.e. if image manager observer method returned `false`, that would mean it is not overridden -> remove images.

Reply:
Observer -> Client :smile:

Reply:
Since Tobrun is OOO, let me throw my two cents that I have no strong preference, and we might as well follow the current observer setup.

On the other hand, could you explain briefly what it takes to mark the image as unused? Does it take into consideration runtime-added layers?

Reply:
Generally, in a scenario where a user sets up all required images ahead of time and then manages layers state (added/removed), if we are removing unused icons by default, this can introduce a breaking change. What do you think?

Taking this into consideration I'd rather not have a default listener and document the behavior better.

Reply:
@LukasPaczos will try my best :rofl: 

**Supershort:** Unused images are images whose requesting tile is destroyed. Core will notify about unused images only if image was provided  thru `onStyleImageMissing(id)` => `addImage(id)`.

> On the other hand, could you explain briefly what it takes to mark the image as unused? 

When tile is parsed / rendered, it might require images that are not present in the style's sprite sheet, thus, core will request image via `onStyleImageMissing` observer API.

If client provides an image, core will associate that image with requesting tile. When tile is destroyed, e.g., source changed, removed or tile pyramid clears tile cache, tile that requested an image will notify image manager that requested image is no longer needed.

Core's renderer will periodically check whether unused images cache is over the limit and notify `MapObserver::onRemoveUnusedStyleImages` that unused resources might be released.

> Does it take into consideration runtime-added layers?

Yes, I don't see why it would not work with runtime added layers / sources.

Reply:
I see, thanks @alexshalamov :smile: I understand that core monitors the `addImage` for the duration of the `onStyleImageMissing` block execution to determine that the image was actually added in that period of time?

Since the `onStyleImageMissing` and `onRemoveUnusedStyleImages` APIs are tightly coupled, we should make sure to reflect that in the javadoc for both of the listeners.

Reply:
:+1: on the approach taken in this PR, as indicated by @LukasPaczos a bit of documentation around this would be great as the comments above show that this behavior isn't hat easy to explain :)

Reply:
Moved implementation to core, we can reconsider that in the future if needed, does not affect public API.